tls/s2n_ktls.c

```diff
@@ -0,0 +1,203 @@
+    #else
+        #include <netinet/tcp.h>
+    #endif
+
+    #include <linux/tls.h>
+#endif
+
+#define S2N_TLS_ULP_NAME      "tls"
+#define S2N_TLS_ULP_NAME_SIZE sizeof(S2N_TLS_ULP_NAME)
+
+#include <sys/socket.h>
+
+#include "tls/s2n_ktls.h"
+#include "utils/s2n_safety.h"
+#include "utils/s2n_socket.h"
+
+/* These variables are used to disable ktls mechanisms during testing. */
+static bool ignore_ktls_ulp_for_testing = false;
+
+S2N_RESULT s2n_ktls_validate(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn->secure);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg->cipher);
+    const struct s2n_cipher *cipher = conn->secure->cipher_suite->record_alg->cipher;
+    RESULT_ENSURE_REF(cipher);
+
+    /* TODO support TLS 1.3
+     *
+     * TLS 1.3 support requires sending the KeyUpdate message when the cryptographic
+     * KeyLimits are met. However, this is currently only possible by applying a
+     * kernel patch to support this functionality.
+     */
+    RESULT_ENSURE(conn->actual_protocol_version == S2N_TLS12, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+
+    /* Check if the cipher supports kTLS */
+    RESULT_ENSURE(cipher->ktls_supported, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_validate_socket_mode(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE(ktls_mode == S2N_KTLS_MODE_RECV || ktls_mode == S2N_KTLS_MODE_SEND, S2N_ERR_SAFETY);
+
+    /* kTLS I/O functionality is managed by s2n-tls. kTLS cannot be enabled if the
+     * application sets custom I/O (managed_send_io == false means application has
+     * set custom I/O).
+     *
+     * - Safety checks
+     * - Confirm application is not using custom I/O
+     * - Confirm kTLS isn't enabled already
+     */
+    switch (ktls_mode) {
+        case S2N_KTLS_MODE_SEND:
+            RESULT_ENSURE(conn->managed_send_io, S2N_ERR_KTLS_SEND_MANAGED_IO);
+            RESULT_ENSURE(!conn->ktls_send_enabled, S2N_ERR_KTLS_SEND_ENABLED);
+            break;
+        case S2N_KTLS_MODE_RECV:
+            RESULT_ENSURE(conn->managed_recv_io, S2N_ERR_KTLS_RECV_MANAGED_IO);
+            RESULT_ENSURE(!conn->ktls_recv_enabled, S2N_ERR_KTLS_RECV_ENABLED);
+            break;
+        case S2N_KTLS_MODE_DUPLEX:
+            RESULT_BAIL(S2N_ERR_SAFETY);
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_ktls_mode ktls_mode, int *fd)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(fd);
+
+    if (ktls_mode == S2N_KTLS_MODE_RECV) {
+        RESULT_ENSURE_REF(conn->recv_io_context);
+        /* retrieve the receive fd */
+        const struct s2n_socket_read_io_context *peer_socket_ctx = conn->recv_io_context;
+        *fd = peer_socket_ctx->fd;
```

Comment:
https://github.com/aws/s2n-tls/pull/3808#discussion_r1103315877

This is now more dangerous, since s2n_ktls_retrieve_file_descriptor doesn't even call s2n_ktls_validate_socket_mode directly anymore. Please do your safety checks in the same method that uses the values.

Reply:
I perform the ref check just above but I think we also need the managed_io checks.

Reply:
Yes, you need the managed_io checks here. Otherwise, you're casting a `void*` to a `struct s2n_socket_read_io_context*` without knowing that's what it actually is.